File: src/lib/dataStructures.ts
export class RingBuffer<T> {
  buffer: T[];
  head: number;
  size: number;

  constructor(size: number = 1024) {
    this.buffer = new Array(size);
    this.head = 0;
    this.size = size;
  }
  push(item: T) {
    this.buffer[this.head] = item;
    this.head = (this.head + 1) % this.size;
  }
  getAll(): T[] {
    return this.buffer.filter(Boolean);
  }
}

export class TemporalBucket {
  count: number = 0;
  sum: number = 0;
  sumSq: number = 0;

  update(price: number) {
    this.count++;
    this.sum += price;
    this.sumSq += price * price;
  }
  getVariance(): number {
    if (this.count < 2) return 0;
    const mean = this.sum / this.count;
    return (this.sumSq / this.count) - (mean * mean);
  }
}

export type TickData = { price: number; ts: number };

export class LockFreeTemporalAggregator {
  stocks: Map<string, { ringBuffer: RingBuffer<TickData>; buckets: Map<number, TemporalBucket> }>;
  priceHistory: Map<string, TickData[]>;

  constructor() {
    this.stocks = new Map();
    this.priceHistory = new Map();
  }

  registerSymbol(symbol: string) {
    if (!this.stocks.has(symbol)) {
      this.stocks.set(symbol, {
        ringBuffer: new RingBuffer(512),
        buckets: new Map(),
      });
      this.priceHistory.set(symbol, []);
    }
  }

  onTick(symbol: string, tick: TickData) {
    this.registerSymbol(symbol);
    const data = this.stocks.get(symbol)!;
    data.ringBuffer.push(tick);
    
    const history = this.priceHistory.get(symbol)!;
    history.push(tick);
    if (history.length > 120) history.shift();

    const bucketKey = Math.floor(tick.ts / 10000); // 10s buckets
    if (!data.buckets.has(bucketKey)) data.buckets.set(bucketKey, new TemporalBucket());
    data.buckets.get(bucketKey)!.update(tick.price);
  }

  getVolatility(symbol: string): number {
    const data = this.stocks.get(symbol);
    if (!data) return 0;
    let totalVar = 0;
    let count = 0;
    data.buckets.forEach((b) => {
      totalVar += b.getVariance();
      count++;
    });
    return Math.sqrt(totalVar / Math.max(count, 1)) || 0.001;
  }
}

File: src/lib/dataStructures.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>
#include <cmath>
#include <chrono>

using namespace std;

// ==========================================
// RING BUFFER
// ==========================================
template <typename T>
class RingBuffer {
private:
    vector<T> buffer;
    int head;
    int tail;
    int capacity;
    int count;

public:
    RingBuffer(int size = 1024) {
        capacity = size;
        buffer.resize(size);
        head = 0;
        tail = 0;
        count = 0;
    }

    void push(T item) {
        buffer[head] = item;
        head = (head + 1) % capacity;
        if (count < capacity) {
            count++;
        } else {
            tail = (tail + 1) % capacity;
        }
    }

    vector<T> getRecent(int n) {
        int clamped = min(n, count);
        vector<T> result;
        for (int i = clamped - 1; i >= 0; i--) {
            int idx = (head - 1 - i + capacity) % capacity;
            result.push_back(buffer[idx]);
        }
        return result;
    }
};

// ==========================================
// TEMPORAL BUCKET
// ==========================================
class TemporalBucket {
public:
    int count = 0;
    double sum = 0.0;
    double sumSq = 0.0;
    long long startTs;
    long long endTs = 0;
    bool isSealed = false;

    // Welford's algorithm trackers
    double wMean = 0.0;
    double wM2 = 0.0;

    TemporalBucket(long long startTs) : startTs(startTs) {}

    void update(double price) {
        if (isSealed) return;
        
        count++;
        sum += price;
        sumSq += price * price;

        double delta = price - wMean;
        wMean += delta / count;
        double delta2 = price - wMean;
        wM2 += delta * delta2;
    }

    double getVariance() {
        if (count < 2) return 0.0;
        double mean = sum / count;
        return max(0.0, (sumSq / count) - (mean * mean));
    }

    double getWelfordVariance() {
        if (count < 2) return 0.0;
        return wM2 / (count - 1);
    }

    void seal(long long end) {
        endTs = end;
        isSealed = true;
    }
};

// ==========================================
// LOCK-FREE TEMPORAL AGGREGATOR
// ==========================================
struct TickData {
    double price;
    long long ts;
};

struct AssetData {
    RingBuffer<TickData> ringBuffer = RingBuffer<TickData>(512);
    unordered_map<long long, TemporalBucket*> buckets;
    long long activeBucketKey = -1;
};

class LockFreeTemporalAggregator {
private:
    unordered_map<string, AssetData*> stocks;
    long long bucketDurationMs;

public:
    LockFreeTemporalAggregator(long long durationMs = 10000) {
        bucketDurationMs = durationMs;
    }

    void registerSymbol(string symbol) {
        if (stocks.find(symbol) == stocks.end()) {
            stocks[symbol] = new AssetData();
        }
    }

    void onTick(string symbol, TickData tick) {
        registerSymbol(symbol);
        AssetData* data = stocks[symbol];

        data->ringBuffer.push(tick);

        long long bucketKey = tick.ts / bucketDurationMs;
        if (data->buckets.find(bucketKey) == data->buckets.end()) {
            if (data->activeBucketKey != -1 && data->buckets.find(data->activeBucketKey) != data->buckets.end()) {
                data->buckets[data->activeBucketKey]->seal(tick.ts);
            }
            data->buckets[bucketKey] = new TemporalBucket(tick.ts);
            data->activeBucketKey = bucketKey;
        }
        data->buckets[bucketKey]->update(tick.price);
    }

    double getVolatility(string symbol) {
        if (stocks.find(symbol) == stocks.end()) return 0.0;
        AssetData* data = stocks[symbol];

        double totalVar = 0.0;
        int count = 0;
        
        for (auto const& [key, bucket] : data->buckets) {
            totalVar += bucket->getWelfordVariance();
            count++;
        }

        if (count == 0) return 0.001;
        return sqrt(totalVar / count);
    }
};
